File: gpio.rb
begin
  require 'rpi_gpio'
rescue RuntimeError
end


class GPIO
  #  7 and 13: forward
  # 11 and 15: backward
  #  7 and 15: left
  # 11 and 13: right
  PINS = [7, 11, 13, 15]
  FREQ = 800

  def initialize
    RPi::GPIO.set_numbering :board
    for pin in PINS
      RPi::GPIO.setup pin, as: :output, initialize: :low
    end
    @pwms = PINS.collect do |pin|
      pwm = RPi::GPIO::PWM.new pin, FREQ
      pwm.start 0
    end
  end

  def update *values
    @pwms.zip(values).each do |pwm, value|
      pwm.duty_cycle = value
    end
  end

  def stop
    @pwms.collect { |pwm| pwm.stop }
    RPi::GPIO.reset
  end
end

File: gpio.py
import signal
try:
    import RPi.GPIO
except RuntimeError:
    pass


class GPIO:
    PINS = [7, 11, 13, 15]
    FREQ = 800

    def __init__(self):
        RPi.GPIO.setmode(RPi.GPIO.BOARD)
        for pin in self.PINS:
            RPi.GPIO.setup(pin, RPi.GPIO.OUT)
        self.pwms = [RPi.GPIO.PWM(pin, self.FREQ) for pin in self.PINS]
        for pwm in self.pwms:
            pwm.start(0)
        signal.signal(signal.SIGTERM, self.stop)
        signal.signal(signal.SIGINT, self.stop)

    def update(self, *values):
        for pwm, value in zip(self.pwms, values):
            pwm.ChangeDutyCycle(value)

    def stop(self):
        for pwm in self.pwms:
            pwm.stop()
        RPi.GPIO.cleanup()
